snapshots/windows/lcow.go

```diff
@@ -0,0 +1,318 @@
+	}
+	m := l.lcowMounts(snapshot)
+	return m, nil
+}
+
+// Remove abandons the transaction identified by key. All resources
+// associated with the key will be removed.
+func (l *lcowSnapshotter) Remove(ctx context.Context, key string) error {
+	ctx, t, err := l.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+
+	id, snInfo, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return errors.Wrapf(errdefs.ErrFailedPrecondition, "failed to get snapshot info: %s", err)
+	}
+
+	_, _, err = storage.Remove(ctx, key)
+	if err != nil {
+		return errors.Wrap(err, "failed to remove")
+	}
+
+	path := l.getSnapshotDir(id)
+	renamed := l.getSnapshotDir("rm-" + id)
+	if err := os.Rename(path, renamed); err != nil && !os.IsNotExist(err) {
+		// Sometimes if there are some open handles to the files (especially VHD)
+		// inside the snapshot directory the rename call will return "access
+		// denied" or "file is being used by another process" errors.  Just
+		// returning that error causes the entire snapshot garbage collection
+		// operation to fail. To avoid that we return failed pre-condition error
+		// here so that snapshot garbage collection can continue and can cleanup
+		// other snapshots.
+		return errors.Wrap(errdefs.ErrFailedPrecondition, err.Error())
+	}
+
+	if err := t.Commit(); err != nil {
+		if err1 := os.Rename(renamed, path); err1 != nil {
+			// May cause inconsistent data on disk
+			log.G(ctx).WithError(err1).WithField("path", renamed).Errorf("Failed to rename after failed commit")
+		}
+		return errors.Wrap(err, "failed to commit")
+	}
+
+	_, hasOverride := snInfo.Labels[labelScratchSnapshotLocation]
+	if hasOverride {
+		rmPath := l.getResolvedSnapshotDir(id, snInfo)
+		if err := os.RemoveAll(rmPath); err != nil {
+			// Must be cleaned up, any "rm-*" could be removed if no active transactions
+			log.G(ctx).WithError(err).WithField("path", rmPath).Warnf("Failed to remove root filesystem")
+		}
+	}
+	if err := os.RemoveAll(renamed); err != nil && !os.IsNotExist(err) {
+		log.G(ctx).WithError(err).Warnf("failed to remove snapshot dir: %s", renamed)
+	}
+
+	return nil
+}
+
+func (l *lcowSnapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts []snapshots.Opt) (_ []mount.Mount, err error) {
+	ctx, t, err := l.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+	defer t.Rollback()
+
+	newSnapshot, snapshotInfo, err := l.createSnapshotCommon(ctx, kind, key, parent, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if kind == snapshots.KindActive {
+		log.G(ctx).Debug("createSnapshot active")
+
+		// Create the new snapshot dir
+		snDir, snOverrideDir, err := l.createSnapshotDirectory(ctx, snapshotInfo, key, newSnapshot.ID)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create snapshot directory")
+		}
+		defer onErrorDirectoryCleanup(ctx, &err, snDir, snOverrideDir)
```

Comment:
since this is the same as in `(*wcowSnapshotter).createSnapshot(`, can this be reorganized to having a function in `windowsSnapshotterBase` perform this setup and then it calls a function handle that `lcowSnapshotter` and `wcowSnapshotter` pass it?

Reply:
I changed this so that the `createSanspshotCommon` function now also handles directory creation. I didn't want to move the `TransactionContext` part in there too because then that function would have to return the transactor which would be used by the caller to either commit or rollback the transaction and I didn't like that idea. I feel that function which initiates the transaction should also create the transactor and should commit or rollback the transaction. Let me know what you think.

Reply:
That makes sense to me